### PR TITLE
Senders recovery stage

### PR DESCRIPTION
**Pull request by jomuel**
Naive implementation of stage 4 of the staged sync, recovery of transaction senders. Tbd: concurrency, configurability, unwind method.

### CODE REVIEW

File: src/stagedsync/stage.rs [hunk 1]
```diff
@@ -48,3 +50,11 @@ pub struct UnwindInput {
     pub stage_progress: u64,
     pub unwind_to: u64,
 }
+
+#[derive(Error, Debug)]
+pub enum SyncError {
```
Comment (vorot93) [inline src/stagedsync/stage.rs:55]: Out of scope for staged sync framework, move into stage

File: src/stages/sender_recovery.rs [hunk 2]
```diff
@@ -0,0 +1,300 @@
...
+        &RecoverableSignature::from_compact(&sig, rec)?,
+    )?;
+
+    let address_slice = &Keccak256::digest(&public.serialize_uncompressed()[1..])[12..];
+    Ok(Address::from_slice(address_slice))
+}
+
+async fn process_block<'tx, RwTx>(tx: &mut RwTx, height: u64) -> anyhow::Result<()>
+where
+    RwTx: MutableTransaction<'tx>,
```
Comment (vorot93) [inline src/stages/sender_recovery.rs:37]:
```suggestion
async fn process_block<'db: 'tx, 'tx, RwTx>(tx: &'tx mut RwTx, height: u64) -> anyhow::Result<()>
where
    RwTx: MutableTransaction<'db>,
```

Maybe stick to original lifetimes, if only for readability?

File: src/stages/sender_recovery.rs [hunk 3]
```diff
@@ -0,0 +1,300 @@
...
+        let mut height = from_height;
+        while height < to_height {
+            if let Err(e) = process_block(tx, height + 1).await {
+                error!(
+                    "Failed to recover senders for block {}: {}",
+                    height + 1,
+                    e.to_string(),
+                );
+                break;
+            }
```
Comment (vorot93) [inline src/stages/sender_recovery.rs:92]:
```suggestion
            process_block(tx, height + 1).await.with_context(|| {
                format!("Failed to recover senders for block {}"
            }))?;
```

Stages must not handle errors, let alone suppress them.

**Review by vorot93 — APPROVED** [review]
LGTM, modulo nits above

File: src/stagedsync/stage.rs (hunk 1 above)
Comment (jomuel) [reply src/stagedsync/stage.rs]: Moved to `sender_recovery.rs`

**Pull request merged**